scripts/imgtool/keys/imgtool_keys_pkcs11.py

```diff
@@ -0,0 +1,179 @@
+        pkcs11_module_path = ''
+        try:
+            pkcs11_module_path = env['PKCS11_MODULE']
+            if not Path(pkcs11_module_path).is_file():
+                raise RuntimeError('$PKCS11_MODULE is not a file: %s' % pkcs11_module_path)
+        except KeyError:
+            pkcs11_module_path = find_library('opensc-pkcs11.so')
+            if pkcs11_module_path is None:
+                raise RuntimeError('$PKCS11_MODULE not set and OpenSC not found.')
+
+        lib = ''
+        try:
+            lib = pkcs11.lib(pkcs11_module_path)
+        except RuntimeError:
+            pass  # happens if lib does not exist or is corrupt
+        if '' == lib:
+            raise RuntimeError('PKCS #11 module not loaded.')
+
+        self.token = lib.get_token(token_serial=params[b'serial'])
+        # try to open a session to see if the PIN is valid
+        with self.token.open(user_pin=self.user_pin) as session:
+            pass
+        self.key_id = params[b'id']
+
+    def shortname(self):
+        return "ecdsa"
+
+    def _unsupported(self, name):
+        raise PKCS11UsageError("Operation {} requires private key".format(name))
+
+    def get_public_bytes(self):
+        with self.token.open(user_pin=self.user_pin) as session:
+            pub = session.get_key(
+                id=self.key_id,
+                key_type=pkcs11.KeyType.EC,
+                object_class=pkcs11.ObjectClass.PUBLIC_KEY
+            )
+            key = oscrypto.asymmetric.load_public_key(
+                pkcs11.util.ec.encode_ec_public_key(pub)
+            )
+            key = pkcs11.util.ec.encode_ec_public_key(pub)
+        return key
+
+    def get_private_bytes(self, minimal):
+        self._unsupported('get_private_bytes')
+
+    def export_private(self, path, passwd=None):
+        self._unsupported('export_private')
+
+    def export_public(self, path):
+        """Write the public key to the given file."""
+        with self.token.open(user_pin=self.user_pin) as session:
+            pub = session.get_key(
+                id=self.key_id,
+                key_type=pkcs11.KeyType.EC,
+                object_class=pkcs11.ObjectClass.PUBLIC_KEY
+            )
+            key = oscrypto.asymmetric.load_public_key(
+                pkcs11.util.ec.encode_ec_public_key(pub)
+            )
+            pem = oscrypto.asymmetric.dump_public_key(key, encoding='pem')
+
+        with open(path, 'wb') as f:
+            f.write(pem)
+
+    def sig_type(self):
+        return "ECDSA256_SHA256"
+
+    def sig_tlv(self):
+        return "ECDSA256"
+
+    def sig_len(self):
+        # The DER encoding depends on the high bit, and can be
+        # anywhere from 70 to 72 bytes.  Because we have to fill in
+        # the length field before computing the signature, however,
+        # we'll give the largest, and the sig checking code will allow
+        # for it to be up to two bytes larger than the actual
+        # signature.
```

Comment:
Since version v1.5.0 MCUboot no longer requires signature padding using zeros.
Copy/paste comment from https://github.com/mcu-tools/mcuboot/blob/main/scripts/imgtool/keys/ecdsa.py

```suggestion
        # Early versions of MCUboot (< v1.5.0) required ECDSA
        # signatures to be padded to 72 bytes.  Because the DER
        # encoding is done with signed integers, the size of the
        # signature will vary depending on whether the high bit is set
        # in each value.  This padding was done in a
        # not-easily-reversible way (by just adding zeros).
        #
        # The signing code no longer requires this padding, and newer
        # versions of MCUboot don't require it.  But, continue to
        # return the total length so that the padding can be done if
        # requested.
```